cmd/thanos/compact.go

```diff
@@ -344,6 +344,13 @@ func runCompact(
 		level.Warn(logger).Log("msg", "Max compaction level is lower than should be", "current", maxCompactionLevel, "default", compactions.maxLevel())
 	}
 
+	// Ensure the lowest resolution downsampling is triggered at the highest compaction level.
+	downsample_range1 := downsample.DownsampleRange1 * time.Millisecond
+	if downsample_range1 <= compactions[len(compactions)-2] ||
+		downsample_range1 >= compactions[len(compactions)-1] {
+		panic("DownsampleRange1 must be between the final two compaction levels")
```

Comment:
We should never ever panic: https://thanos.io/contributing/coding-style-guide.md/#never-use-panics

What we are trying to achieve here? `compactions` are hardcoded so we can totally test this on unit test, no?

Reply:
This code is asserting an invariant between two internal constants.  It's a developer-facing panic, users will never see it.

The program should not run if these constants are misconfigured.  Using a runtime error is a stronger safeguard against that than a unit test.

Thanos codebase uses panic elsewhere, e.g. https://github.com/thanos-io/thanos/blob/2634e0331487248becb0b1e7c49f88aa533e2bb7/cmd/thanos/main.go#L86